### PR TITLE
ARP: implement decoder and logger v5

### PR DESCRIPTION
**Pull request by glongo**
Make sure these boxes are signed before submitting your Pull Request -- thank you.

- [x] I have read the contributing guide lines at
   https://docs.suricata.io/en/latest/devguide/contributing/contribution-process.html
- [x] I have signed the Open Information Security Foundation contribution agreement at
   https://suricata.io/about/contribution-agreement/ (note: this is only required once)
- [x] I have updated the user guide (in doc/userguide/) to reflect the changes made (if applicable)

Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) ticket:
https://redmine.openinfosecfoundation.org/issues/6827

Describe changes:
- ckeck 5-tuple values prior to logging

The entry now appears as follows:

```
{
  "timestamp": "2012-11-11T02:00:25.652633+0100",
  "pcap_cnt": 5,
  "event_type": "arp",
  "pkt_src": "wire/pcap",
  "arp": {
    "hw_type": "ethernet",
    "proto_type": "ipv4",
    "opcode": "request",
    "src_mac": "00:1d:09:f0:92:ab",
    "src_ip": "10.10.10.1",
    "dest_mac": "00:00:00:00:00:00",
    "dest_ip": "10.10.10.2"
  }
}
```

SV_BRANCH=https://github.com/OISF/suricata-verify/pull/1739

### CODE REVIEW

File: src/output-json.c
Comment (jasonish) [inline src/output-json.c:845]: Should we worry about scanning the string on almost every log entry? I think we could just check for the first byte being != '\0'?

**Review by jasonish — COMMENTED** [review]
Looking good, I'm running with it out. Already seeing one of my tools break as the expectation of an `eve` record other than stats always have IP addresses, but that's not a concern of Suricata's.

File: src/output-json.c
Comment (victorjulien) [reply src/output-json.c:845]: I think we can count on `strlen` being optimized by the library.

**Comment by glongo** [comment]
new SV_BRANCH=https://github.com/OISF/suricata-verify/pull/1739

File: src/output-json.c
Comment (catenacyber) [reply src/output-json.c:845]: > I think we can count on strlen being optimized by the library.

I would not bet on that

File: src/output-json-arp.c
Comment (catenacyber) [inline src/output-json-arp.c:47]: Do we really need all these includes ? For instance, there are no unit tests in this file...

**Review by catenacyber — CHANGES_REQUESTED** [review]
Will at least need a rebase

**Comment by victorjulien** [comment]
The work in #10796 was triggered by this work, as I didn't like adding a new pointer to the packet. So my request is to wait until that work is merged, then rebase while implementing the new pattern introduced there.

**Comment by victorjulien** [comment]
This can now be rebased. You will have to update this new L3 section
```c
  enum PacketL3Types {
      PACKET_L3_UNKNOWN = 0,
      PACKET_L3_IPV4,
      PACKET_L3_IPV6,
  };
  
  struct PacketL3 {
      enum PacketL3Types type;
      /* Checksum for IP packets. */
      bool csum_set;
      uint16_t csum;
      union Hdrs {
          IPV4Hdr *ip4h;
          IPV6Hdr *ip6h;
      } hdrs;
      /* IPv4 and IPv6 are mutually exclusive */
      union {
          IPV4Vars ip4;
          struct {
              IPV6Vars v;
              IPV6ExtHdrs eh;
          } ip6;
      } vars;
  };
```
And refactor macro's to use the new style inline functions.

**Comment by glongo** [comment]
> This can now be rebased. You will have to update this new L3 section
> 
> ```c
>   enum PacketL3Types {
>       PACKET_L3_UNKNOWN = 0,
>       PACKET_L3_IPV4,
>       PACKET_L3_IPV6,
>   };
>   
>   struct PacketL3 {
>       enum PacketL3Types type;
>       /* Checksum for IP packets. */
>       bool csum_set;
>       uint16_t csum;
>       union Hdrs {
>           IPV4Hdr *ip4h;
>           IPV6Hdr *ip6h;
>       } hdrs;
>       /* IPv4 and IPv6 are mutually exclusive */
>       union {
>           IPV4Vars ip4;
>           struct {
>               IPV6Vars v;
>               IPV6ExtHdrs eh;
>           } ip6;
>       } vars;
>   };
> ```
> 
> And refactor macro's to use the new style inline functions.

Wouldn't it be better to update the L2 struct? That seems more appropriate to me.

**Comment by victorjulien** [comment]
> > This can now be rebased. You will have to update this new L3 section
> > ```c
> >   enum PacketL3Types {
> >       PACKET_L3_UNKNOWN = 0,
> >       PACKET_L3_IPV4,
> >       PACKET_L3_IPV6,
> >   };
> >   
> >   struct PacketL3 {
> >       enum PacketL3Types type;
> >       /* Checksum for IP packets. */
> >       bool csum_set;
> >       uint16_t csum;
> >       union Hdrs {
> >           IPV4Hdr *ip4h;
> >           IPV6Hdr *ip6h;
> >       } hdrs;
> >       /* IPv4 and IPv6 are mutually exclusive */
> >       union {
> >           IPV4Vars ip4;
> >           struct {
> >               IPV6Vars v;
> >               IPV6ExtHdrs eh;
> >           } ip6;
> >       } vars;
> >   };
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > And refactor macro's to use the new style inline functions.
> 
> Wouldn't it be better to update the L2 struct? That seems more appropriate to me.

ARP is a layer on top of the L2 Ethernet layer, so for us it would fit L3.

**Pull request closed without merging**
